### PR TITLE
syncer/: fix flush checkpoint for ddls for situations that IsSharding is false

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
https://github.com/pingcap/dm/issues/514
* When `IsSharding` is false, table checkpoint which contains schema tracker info was not updated in `dm_meta` instantly. But global checkpoint is updated. If worker fails now, it will cause a mismatch between tracker and global checkpoint.
* Fix a dead lock bug.

### What is changed and how it works?
Update table point when we update global checkpoint.

### Check List 

Tests 

 - Unit test
 - Integration test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by lichunzhu** [comment]
Should be merged after https://github.com/pingcap/dm/pull/581/files, add uncomment the blocked integration tests

File: syncer/syncer.go [hunk 1]
```diff
@@ -742,15 +742,19 @@ func (s *Syncer) addJob(job *job) error {
...
 		}
 		// reset sharding group after checkpoint saved
 		s.resetShardingGroup(job.targetSchema, job.targetTable)
 	case insert, update, del:
 		// save job's current pos for DML events
-		if len(job.sourceSchema) > 0 {
-			s.saveTablePoint(job.sourceSchema, job.sourceTable, job.currentLocation)
+		for i := range job.sourceSchemas {
+			if len(job.sourceSchemas[i]) > 0 {
+				s.saveTablePoint(job.sourceSchemas[i], job.sourceTables[i], job.location)
```
Comment (csuzhangxc) [inline syncer/syncer.go:756]: Should we use `job. location` or `job.currentLocation` here?

**Comment by csuzhangxc** [comment]
in #514, I think we should also **flush** (not only **save**) checkpoint **before** executing DDL to solve a case:

1. `INSET TABLE tbl (c1) VALUES (1)`
2. `ALTER TABLE tbl ADD COLUMN (c2 INT)`
3. flush checkpoint into downstream TiDB

interrupted after step.2 and before step.3, and then restarted from step.1, this will trigger an error with `columns and data mismatch in length: 2 (columns) 1 (data)`

File: syncer/syncer.go (hunk 1 above)
Comment (lichunzhu) [reply syncer/syncer.go]: updated in https://github.com/pingcap/dm/pull/587/commits/f1837f8d8dd5d9aba2c9442fc5f5f9326745a019

**Comment by lichunzhu** [comment]
> in #514, I think we should also **flush** (not only **save**) checkpoint **before** executing (and track) DDL to solve a case:
> 
> 1. `INSET TABLE tbl (c1) VALUES (1)`
> 2. `ALTER TABLE tbl ADD COLUMN (c2 INT)`
> 3. flush checkpoint into downstream TiDB
> 
> interrupted after step.2 and before step.3, and then restarted from step.1, this will trigger an error with `columns and data mismatch in length: 2 (columns) 1 (data)`

updated in https://github.com/pingcap/dm/pull/587/commits/f1837f8d8dd5d9aba2c9442fc5f5f9326745a019. A test is also added in this commit.

File: syncer/optimist.go [hunk 2]
```diff
@@ -78,6 +78,10 @@ func (s *Syncer) handleQueryEventOptimistic(
 		}
 	}
 
+	if err := s.flushCheckPoints(); err != nil {
```
Comment (csuzhangxc) [inline syncer/optimist.go:81]: Can we put it near `s.jobWg.Wait()` to make it more meaningful?

File: syncer/syncer.go [hunk 3]
```diff
@@ -753,11 +753,16 @@ func (s *Syncer) addJob(job *job) error {
 		// save job's current pos for DML events
 		for i := range job.sourceSchemas {
 			if len(job.sourceSchemas[i]) > 0 {
-				s.saveTablePoint(job.sourceSchemas[i], job.sourceTables[i], job.location)
+				s.saveTablePoint(job.sourceSchemas[i], job.sourceTables[i], job.currentLocation)
 			}
 		}
 	}
 
+	failpoint.Inject("FlushCheckpointExit", func() {
```
Comment (csuzhangxc) [inline syncer/syncer.go:761]: in many places, we call `s.flushCheckPoints` directly, but they will not trigger this `FlushCheckpointExit`, how about giving it another name or move it into `s.flushCheckPoints`?

File: syncer/syncer.go [hunk 4]
```diff
@@ -1706,6 +1711,10 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext) e
 			needHandleDDLs = appliedSQLs // maybe nil
 		}
 
+		if err := s.flushCheckPoints(); err != nil {
```
Comment (csuzhangxc) [inline syncer/syncer.go:1714]: If we do not dispatch a `flush` job, then we must call `s.jobWg.Wait()` before `s.flushCheckPoints()`.

(Is dispatch a `flush` job better?)

File: tests/sharding2/run.sh [hunk 5]
```diff
@@ -46,6 +46,40 @@ function run() {
     # use sync_diff_inspector to check data now!
     echo "check sync diff for the increment replication"
     check_sync_diff $WORK_DIR $cur/conf/diff_config.toml
+
+    echo "kill dm-worker1"
+    ps aux | grep dm-worker1 |awk '{print $2}'|xargs kill || true
+    echo "kill dm-worker2"
+    ps aux | grep dm-worker2 |awk '{print $2}'|xargs kill || true
```
Comment (csuzhangxc) [inline tests/sharding2/run.sh:53]: Do we need to `check_port_offline` here?

File: syncer/syncer.go [hunk 6]
```diff
@@ -1799,7 +1798,7 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext) e
 
 	s.tctx.L().Info(annotate, zap.String("event", "query"), zap.String("source", source), zap.Strings("ddls", needHandleDDLs), zap.ByteString("raw statement", ev.Query), zap.Bool("in-sharding", needShardingHandle), zap.Stringer("start location", startLocation), zap.Bool("is-synced", synced), zap.Int("unsynced", remain))
 
-	if err := s.flushCheckPoints(); err != nil {
+	if err := s.flushJobs(); err != nil {
```
Comment (csuzhangxc) [inline syncer/syncer.go:1801]: It seems we can remove `s.jobWg.Wait()` in L1856 now? (or even we can remove `s.jobWg.Wait()` in L723? but we may still just keep it, the shard mechanism make this not uniform, 😢 )

File: syncer/job.go [hunk 7]
```diff
@@ -58,8 +58,8 @@ func (t opType) String() string {
 
 type job struct {
 	tp              opType
-	sourceSchema    string
-	sourceTable     string
+	sourceSchemas   []string
+	sourceTables    []string
```
Comment (WangXiangUSTC) [inline syncer/job.go:62]: why job need to use multiple schemas and tables? can you add a comment

File: syncer/job.go [hunk 8]
```diff
@@ -111,10 +112,21 @@ func newDDLJob(ddlInfo *shardingDDLInfo, ddls []string, location, cmdLocation bi
 	}
 
 	if ddlInfo != nil {
-		j.sourceSchema = ddlInfo.tableNames[0][0].Schema
-		j.sourceTable = ddlInfo.tableNames[0][0].Name
+		j.sourceSchemas = []string{ddlInfo.tableNames[0][0].Schema}
+		j.sourceTables = []string{ddlInfo.tableNames[0][0].Name}
 		j.targetSchema = ddlInfo.tableNames[1][0].Schema
 		j.targetTable = ddlInfo.tableNames[1][0].Name
+	} else if sourceTbls != nil {
```
Comment (WangXiangUSTC) [inline syncer/job.go:119]: in which case `ddlInfo` is nil and `sourceTbls` is not nil? can you add a example

File: syncer/job.go (hunk 7 above)
Comment (WangXiangUSTC) [reply syncer/job.go:62]: and I think map is better

File: syncer/job.go (hunk 8 above)
Comment (lichunzhu) [reply syncer/job.go:122]: updated in https://github.com/pingcap/dm/pull/587/commits/2733185c2787894f9104e20aab935ac154092e4b

File: syncer/job.go (hunk 7 above)
Comment (lichunzhu) [reply syncer/job.go]: all updated in https://github.com/pingcap/dm/pull/587/commits/2733185c2787894f9104e20aab935ac154092e4b

**Pull request merged**